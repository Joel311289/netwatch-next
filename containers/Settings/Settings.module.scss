/** Container (main component) **/

.wrapper {
  display: block;
}

.section {
  width: 100%;
  margin-bottom: 50px;
}

.heading {
  @include font($size: 14px, $weight: 500);

  text-transform: uppercase;
}

.content {
  display: flex;
}

/** SettingsMode (child component) **/

.settingsMode {
  width: 100% !important;
}

.settingsModeItem {
  flex-grow: 1;
}

.modeIcon {
  @include font($size: 20px);

  margin-top: 2px;

  &.auto {
    @include font($size: 24px);

    margin-top: 5px;
  }
}

.modeLabel {
  @include font($size: 14px, $weight: 500);
}
