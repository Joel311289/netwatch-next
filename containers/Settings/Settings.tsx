import { IconType } from 'react-icons';
import { MdOutlineDarkMode, MdOutlineLightMode, MdSettingsBrightness } from 'react-icons/md';
import { useTheme } from '@/hooks/useTheme';
import { Space } from '@/components/layout';
import { ToggleButton } from '@/components/forms';
import { classes } from '@/utils/helpers';
import styles from '@/containers/Settings/Settings.module.scss';

/* -------------------------------------------------------------------------- */
/** SettingsMode (child component) **/
/* -------------------------------------------------------------------------- */

type SettingMode = {
  key: string;
  label: string;
  Icon: IconType;
};

type SettingsModeProps = {
  modes?: Array<SettingMode>;
};

export const defaultModes: Array<SettingMode> = [
  {
    key: 'light',
    label: 'settings.aspect.option.light',
    Icon: MdOutlineLightMode
  },
  {
    key: 'auto',
    label: 'settings.aspect.option.auto',
    Icon: MdSettingsBrightness
  },
  {
    key: 'dark',
    label: 'settings.aspect.option.dark',
    Icon: MdOutlineDarkMode
  }
];

const SettingsMode = ({ modes = defaultModes }: SettingsModeProps) => {
  const { themeKey, onChangeTheme } = useTheme();

  return (
    <ToggleButton className={styles.settingsMode} activeButton={themeKey} onChange={onChangeTheme}>
      {modes.map(({ key, label, Icon }) => (
        <ToggleButton.Item key={key} value={key} className={styles.settingsModeItem}>
          <Space align="center" gap={8}>
            <div className={classes(styles.modeIcon, styles[key])}>
              <Icon />
            </div>
            <span className={styles.modeLabel}>{label}</span>
          </Space>
        </ToggleButton.Item>
      ))}
    </ToggleButton>
  );
};

/* -------------------------------------------------------------------------- */
/** Container (main component) **/
/* -------------------------------------------------------------------------- */

type SettingSectionContent = ({ modes }: SettingsModeProps) => JSX.Element;

export type SettingSection = {
  key: string;
  heading: string;
  content: SettingSectionContent;
};

type SettingsProps = typeof defaultProps & {
  sections?: Array<SettingSection>;
};

export const defaultProps = {
  sections: [
    {
      key: 'mode',
      heading: 'settings.aspect.heading',
      content: SettingsMode
    }
  ]
};

const Settings = ({ sections }: SettingsProps) => {
  return (
    <Space direction="column" gap={20} className={classes(styles.wrapper)}>
      {sections.map(({ heading, content: Content }) => {
        return (
          <Space key={heading} direction="column" gap={10} className={styles.section}>
            {heading && <span className={styles.heading}>{heading}</span>}

            <div className={styles.content}>
              <Content />
            </div>
          </Space>
        );
      })}
    </Space>
  );
};

Settings.defaultProps = defaultProps;

export default Settings;
