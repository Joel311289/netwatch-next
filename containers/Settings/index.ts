export { default as Settings } from '@/containers/Settings/Settings';
