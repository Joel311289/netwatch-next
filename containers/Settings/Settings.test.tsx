import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Settings, {
  defaultModes,
  defaultProps,
  SettingSection
} from '@/containers/Settings/Settings';

const mockOnChangeThemeValue = jest.fn();

jest.mock('@/hooks/useTheme', () => ({
  useTheme: jest.fn().mockImplementation(() => ({
    theme: 'light',
    darkMode: false,
    onChangeTheme: mockOnChangeThemeValue
  }))
}));

describe('Tests Settings component', () => {
  test('renders container correctly', () => {
    const { container } = render(<Settings />);

    expect(container).toMatchSnapshot();
  });

  describe('Tests SettingsMode ', () => {
    const mode = defaultProps.sections.find(({ key }) => key === 'mode') as SettingSection;
    const defaultModeLabels = defaultModes.map(({ label }) => label).join('|');

    test('renders mode settings correctly', () => {
      render(<Settings />);

      expect(screen.getByText(mode.heading)).toBeInTheDocument();
      expect(screen.getAllByText(new RegExp(`${defaultModeLabels}`, 'i'))).toHaveLength(3);
    });

    defaultModes.forEach(({ key, label }) => {
      test(`changes the ${key} mode from settings correctly`, async () => {
        render(<Settings />);

        userEvent.click(screen.getByText(label));

        await waitFor(() => {
          expect(mockOnChangeThemeValue).toHaveBeenCalledWith(key);
        });
      });
    });
  });
});
