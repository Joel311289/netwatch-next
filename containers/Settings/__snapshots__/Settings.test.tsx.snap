// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Settings component renders container correctly 1`] = `
<div>
  <div
    class="wrapper wrapper"
  >
    <div
      class="wrapper section"
    >
      <span
        class="heading"
      >
        settings.aspect.heading
      </span>
      <div
        class="content"
      >
        <div
          class="wrapper wrapper settingsMode"
        >
          <button
            aria-pressed="false"
            class="itemWrapper settingsModeItem"
          >
            <div
              class="wrapper"
            >
              <div
                class="modeIcon light"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M12 9c1.65 0 3 1.35 3 3s-1.35 3-3 3-3-1.35-3-3 1.35-3 3-3m0-2c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37a.996.996 0 00-1.41 0 .996.996 0 000 1.41l1.06 1.06c.39.39 1.03.39 1.41 0a.996.996 0 000-1.41l-1.06-1.06zm1.06-10.96a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36a.996.996 0 000-1.41.996.996 0 00-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
                  />
                </svg>
              </div>
              <span
                class="modeLabel"
              >
                settings.aspect.option.light
              </span>
              <style>
                
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          gap: 8px;
          
          align-items: center;
        }
      
              </style>
            </div>
          </button>
          <button
            aria-pressed="false"
            class="itemWrapper settingsModeItem"
          >
            <div
              class="wrapper"
            >
              <div
                class="modeIcon auto"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M21 3H3c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16.01H3V4.99h18v14.02zM8 16h2.5l1.5 1.5 1.5-1.5H16v-2.5l1.5-1.5-1.5-1.5V8h-2.5L12 6.5 10.5 8H8v2.5L6.5 12 8 13.5V16zm4-7c1.66 0 3 1.34 3 3s-1.34 3-3 3V9z"
                  />
                </svg>
              </div>
              <span
                class="modeLabel"
              >
                settings.aspect.option.auto
              </span>
              <style>
                
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          gap: 8px;
          
          align-items: center;
        }
      
              </style>
            </div>
          </button>
          <button
            aria-pressed="false"
            class="itemWrapper settingsModeItem"
          >
            <div
              class="wrapper"
            >
              <div
                class="modeIcon dark"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M9.37 5.51A7.35 7.35 0 009.1 7.5c0 4.08 3.32 7.4 7.4 7.4.68 0 1.35-.09 1.99-.27A7.014 7.014 0 0112 19c-3.86 0-7-3.14-7-7 0-2.93 1.81-5.45 4.37-6.49zM12 3a9 9 0 109 9c0-.46-.04-.92-.1-1.36a5.389 5.389 0 01-4.4 2.26 5.403 5.403 0 01-3.14-9.8c-.44-.06-.9-.1-1.36-.1z"
                  />
                </svg>
              </div>
              <span
                class="modeLabel"
              >
                settings.aspect.option.dark
              </span>
              <style>
                
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          gap: 8px;
          
          align-items: center;
        }
      
              </style>
            </div>
          </button>
          <style>
            
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          
          
          align-items: center;
        }
      
          </style>
        </div>
      </div>
      <style>
        
        .wrapper {
          display: flex;
          flex-direction: column;
          flex-wrap: nowrap;
          gap: 10px;
          
          
        }
      
      </style>
    </div>
    <style>
      
        .wrapper {
          display: flex;
          flex-direction: column;
          flex-wrap: nowrap;
          gap: 20px;
          
          
        }
      
    </style>
  </div>
</div>
`;
